x/operator/keeper/consensus_keys.go

```diff
@@ -438,3 +443,114 @@ func (k Keeper) ClearPreviousConsensusKeys(ctx sdk.Context, chainID string) {
+	store := ctx.KVStore(k.storeKey)
+	for i := range prevConsKeys {
+		prevKey := prevConsKeys[i]
+		keys, err := assetstypes.ParseJoinedStoreKey([]byte(prevKey.Key), 2)
+		if err != nil {
+			return err
+		}
+		chainID := keys[0]
+		opAccAddr, err := sdk.AccAddressFromBech32(keys[1])
+		if err != nil {
+			return err
+		}
+		wrappedKey := types.NewWrappedConsKeyFromHex(prevKey.ConsensusKey)
+		bz := k.cdc.MustMarshal(wrappedKey.ToTmProtoKey())
+
+		store.Set(types.KeyForChainIDAndOperatorToPrevConsKey(chainID, opAccAddr), bz)
+	}
+	return nil
+}
+
+func (k *Keeper) GetAllPrevConsKeys(ctx sdk.Context) ([]types.PrevConsKey, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{types.BytePrefixForOperatorAndChainIDToPrevConsKey})
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]types.PrevConsKey, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		var consKey tmprotocrypto.PublicKey
+		k.cdc.MustUnmarshal(iterator.Value(), &consKey)
+		chainID, operatorAddr, err := types.ParsePrevConsKey(iterator.Key())
+		if err != nil {
+			return nil, err
+		}
+		wrappedConsKey := types.NewWrappedConsKeyFromTmProtoKey(&consKey)
+		ret = append(ret, types.PrevConsKey{
+			Key:          string(assetstypes.GetJoinedStoreKey(chainID, operatorAddr.String())),
+			ConsensusKey: wrappedConsKey.ToHex(),
+		})
+	}
+	return ret, nil
+}
+
+func (k *Keeper) SetAllOperatorKeyRemovals(ctx sdk.Context, operatorKeyRemoval []types.OperatorKeyRemoval) error {
+	store := ctx.KVStore(k.storeKey)
+	for i := range operatorKeyRemoval {
+		keyRemoval := operatorKeyRemoval[i]
+		keys, err := assetstypes.ParseJoinedStoreKey([]byte(keyRemoval.Key), 2)
+		if err != nil {
+			return err
+		}
+		chainID := keys[1]
+		opAccAddr, err := sdk.AccAddressFromBech32(keys[0])
+		if err != nil {
+			return err
+		}
+		store.Set(types.KeyForOperatorKeyRemovalForChainID(opAccAddr, chainID), []byte{})
+	}
+	return nil
+}
+
+func (k *Keeper) GetAllOperatorKeyRemovals(ctx sdk.Context) ([]types.OperatorKeyRemoval, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{types.BytePrefixForOperatorKeyRemovalForChainID})
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]types.OperatorKeyRemoval, 0)
+	for ; iterator.Valid(); iterator.Next() {
+		operatorAddr, chainID, err := types.ParseKeyForOperatorKeyRemoval(iterator.Key())
+		if err != nil {
+			return nil, err
+		}
+
+		ret = append(ret, types.OperatorKeyRemoval{
+			Key: string(assetstypes.GetJoinedStoreKey(operatorAddr.String(), chainID)),
+		})
+	}
+	return ret, nil
+}
+
+func (k *Keeper) GetAllOperatorConsKeyRecords(ctx sdk.Context) ([]types.OperatorConsKeyRecord, error) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte{types.BytePrefixForOperatorAndChainIDToConsKey})
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	ret := make([]types.OperatorConsKeyRecord, 0)
```

Comment:
I observed that the memory allocation for similar **slices** is currently set to **zero**. While this approach functions correctly, it is recommended estimating a capacity for these allocations to avoid reallocation each time.

Reply:
we don't know the capacity of these slices, and Golang will allocate a page for the slice, there won't be much of an issue.